### PR TITLE
Feature: Spend for Additional Points Token Option

### PR DESCRIPTION
**Pull request by epsilonError**
### Description

A new type of token option (spend for additional points) is implemented in this pull request. This token option serves as a way to add additional points to a Canvas assignment / quiz by spending tokens. Along with the standard configuration for all token options, the following configurations are available for this type of token option:

1. Canvas Assignment / Quiz to add points to (restricted to those with a `percent` or `points` Canvas grade type)
2. Score to add to the assignment (supplied as a number, `2.5`, or percentage, `10.5%`)
3. Optionally choose an Canvas Assignment Group to base the denominator of additional points on
4. Optionally allow multiple approved requests (-1 for unlimited)
5. Exclude other token options

A student's request will only be approved if all of the following conditions are met:

1. The student does not reach the maximum allowed amount of approved requests for this token option. If the amount of multiple requests is unlimited (-1), this condition will always be met.
2. The student does not have any approved request for token options that this token option is excluded with.
3. If the token balance change for this token option is negative, the student must have sufficient tokens (i.e., have a non-negative balance after spending for this token option)

Token ATM will throw an error and not process student requests if:

- The grading type for the assignment / quiz is not `points` or `percentage` at the time of request processing
- The automatically collected submission score to add points to is does not match the current assignment / quiz score on Canvas

When approved, Token ATM will update the grade of the configured assignment / quiz (including an submission comment of the change executed), record the request, and make the corresponding token balance change.

#### How Grading is Handled Note

The updated grade posted to Canvas can be either a number or percentage, which can be different from the type given as the "Score to add" or the assignment / quiz grade type. Token ATM attempts to chose the most precise representation to update on Canvas. I.e, adding 1 to an assignment currently scored at 50% of 30, will update the grade to 16 not 53.333...%. The grading submission comment attempts to provide the necessary info to check that this math was done correctly.

- Assignments / Quizzes without submissions are treated as 0 scored and then added to.
- The second line of the grading submission comment is given in terms of the data provided in the token option configuration and assignment / quiz info from Canvas.
- The third line of the grading submission comment provides the actual updated grade posted to Canvas, and the original value of the same type that Canvas provided which was added to.

### Testing Note

1. Please test if the new type of token option works as described.
2. Please look over `canvas-grading.spec.ts` for any glaring untested edge-cases. The `canvas-grading.ts` utility provides the code for identifying, parsing, handling, and calculating what should be the updated posted grade that is provided to Canvas.

### CODE REVIEW

**Review by yongkanm — CHANGES_REQUESTED** [review]
### Available configurations to be tested:

- [x] Canvas Assignment / Quiz to add points to (restricted to those with a percent or points Canvas grade type)
- [x] Score to add to the assignment (supplied as a number, 2.5, or percentage, 10.5%)
    
    <del>The score to add cannot be set to a negative value. It makes sense in terms of spending tokens to make grade better, but the error message does not indicate that a negative value is not allowed.</del> Issue is fixed in latest updates.

- [x] Optionally choose an Canvas Assignment Group to base the denominator of additional points on
- [x] Optionally allow multiple approved requests (-1 for unlimited)
- [x] Exclude other token options

### Approval conditions to be tested:

- [x] The student does not reach the maximum allowed amount of approved requests for this token option. If the amount of multiple requests is unlimited (-1), this condition will always be met.
- [x] The student does not have any approved request for token options that this token option is excluded with.
- [x] If the token balance change for this token option is negative, the student must have sufficient tokens (i.e., have a non-negative balance after spending for this token option)

### Error conditions to be tested:

- [x] The grading type for the assignment / quiz is not points or percentage at the time of request processing
- [x] The automatically collected submission score to add points to is does not match the current assignment / quiz score on Canvas

    This works as expected, but I'm wondering if we should throw an error or not in this case. Maybe we need to confirm with the user.

### Request handling behaviors to be tested:

- [x] Assignments / Quizzes without submissions are treated as 0 scored and then added to.
- [x] The second line of the grading submission comment is given in terms of the data provided in the token option configuration and assignment / quiz info from Canvas.

    * <del>The description could be misleading when scroes are added as points. For example, here is a message generated in such as setup: "Added 1 out of 200 total points to 20% of 10 points." The total points of 200 have nothing to do with the actual change. Therefore, I suggest to only show the sum of points possible for the assignment group when the added score is in the form of percentage.</del> Issue is fixed in latest updates.
    * <del>When an assignment group with 0 possible points in total is selected, the assignment group info is omitted from the description, which could also be misleading. For example, when the "Score to Add" is configured to 50%, the message will be "Added 50% to 102 out of 10 points. Change: 102 => 102" This message indicates an addition of 5 points, but the actual change is 0 points due to the assignment group with 0 possible points in total is configured.</del> Issue is fixed in latest updates.
- [x] The third line of the grading submission comment provides the actual updated grade posted to Canvas, and the original value of the same type that Canvas provided which was added to.</del> Issue is fixed in latest updates.

### Note

The following setups have been tested:

1. The assignment grading type as percentage/points
2. "Score to add" as percentage/points
3. Whether specify an assignment group to base the denominator of additional points or not

Due to time limit, `canvas-grading.ts` and `canvas-grading.spec.ts` are not comprehensively reviewed. The conversion between percentage and point behave as expected when performing testings mentioned above. In my opinion, the comphrensive review could be postponed to a time after the approval of this pull request.

<del>A bug is discovered during the testing. In `optional-field.component.ts` at line 28, the null check for `cachedSrcValue` should be changed to `cachedSrcValue !== undefined` to prevent setting `srcValue` as `0` from being skipped.</del> Issue is fixed in latest updates.

Several comments are made. I've resolved comments that does not need to be handled before the pull request is merged.

File: src/app/request-handlers/spend-for-additional-points-request-handler.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    ): Promise<ProcessedRequest> {
+        const { gradingType, pointsPossible } = await this.canvasService.getAssignmentGradingTypeAndPointsPossible(
+            configuration.course.id,
+            request.tokenOption.assignmentId
+        );
+        if (gradingType !== 'points' && gradingType !== 'percent') {
+            throw new Error(
+                `${request.tokenOption.assignmentName} doesn’t display its Grade as Points or as a Percent, and so its score cannot be added to.`
+            );
+        }
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/request-handlers/spend-for-additional-points-request-handler.ts]: I agree that checking for the grading type and publish state of assignment should be checked, but I believe it more belongs to the review functionality rather than checking it for every request. Currently for other token option types, the check is still done passively (i.e., if an API call returns an error, then somethings is wrong).

[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/request-handlers/spend-for-additional-points-request-handler.ts]: But you're right that grading type check is necessary in this case given that posting a grade with invalid value will not get an error but instead make the grade become `null`. I think the responsibility to warn the instructor that something goes wrong should belong to the review functionality, but we definitely need a check for grading type before the API call to change the grade.

File: src/app/request-handlers/spend-for-additional-points-request-handler.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                request.tokenOption.assignmentId,
+                studentRecord.student.id
+            );
+            const totalPointsPossible = request.tokenOption.changeMaxPossiblePoints
+                ? await this.canvasService.getTotalPointsPossibleInAnAssignmentGroup(
+                      configuration.course.id,
+                      request.tokenOption.changeMaxPossiblePoints[1].groupId
+                  )
+                : undefined;
+            // TODO: Handle / find the actual current Submission, or add to all submissions
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/request-handlers/spend-for-additional-points-request-handler.ts:66]: Yeah, Quiz Submission allows you to specify attempt count to get the quiz version for that specific attempt, but Submission is not that flexible.

File: src/app/services/canvas.service.ts [hunk 3]
```diff
@@ -431,6 +433,9 @@ export class CanvasService {
         });
     }
 
+    /**
+     * Warning! Has no internal guards to protect assignments/submissions.
+     */
     public async gradeSubmissionWithPercentage(
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/services/canvas.service.ts:439]: Agreed. I think this should be an optional functionality that an instructor could choose to turn on.

File: src/app/services/canvas.service.ts [hunk 4]
```diff
@@ -1549,4 +1577,78 @@ export class CanvasService {
...
+        assignmentId: string
+    ): Promise<keyof typeof CanvasGradingType> {
+        const data = await this.apiRequest(`/api/v1/courses/${courseId}/assignments/${assignmentId}`);
+        if (typeof data['grading_type'] != 'string') {
+            throw new Error('Invalid data');
+        }
+        return data['grading_type'] as keyof typeof CanvasGradingType;
+    }
+
+    public async getAssignmentGradingTypeAndPointsPossible(
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/services/canvas.service.ts:1592]: We could (and should) always make attributes that does not guaranteed to exist as optional attributes (via `t.union([t.null, t. undefined, <actual type>])`). I'm suggesting this since every REST API call will return these attributes regardless whether you need it or not, but things might change with GraphQL.

File: src/app/utils/canvas-grading.ts [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+    const secondLine = `Change: ${postedGrade.includes('%') ? target.grade : target.score} => ${postedGrade}`;
+    return firstLine + secondLine;
+}
+
+/**
+ * @param postedGrade The score or grade posted on canvas as a string.
+ * @returns The parsed number in decimals, or NaN if not correctly parsable.
+ */
+export function parseCanvasPercentsAndPoints(postedGrade: string): number {
+    const checkNonNumber = /[^\d.%]/m;
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/utils/canvas-grading.ts]: Issue is fixed in latest updates.

File: src/app/request-handlers/spend-for-additional-points-request-handler.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            if (!gradeMatchesCurrentSubmission) {
+                throw new Error('Was about to add to an out-of-date Canvas submission score.');
+            }
+            const { postedGrade: newPostedGrade, updateMessage } = addPercentOrPointsToCanvasGrade(
+                addText,
+                { gradeType: gradingType, grade, score, pointsPossible },
+                totalPointsPossible
+            );
+            const assignmentComment = `Request for ${request.tokenOption.name} was approved.\n` + updateMessage;
+            this.canvasService.postSubmissionGradeWithComment(
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/request-handlers/spend-for-additional-points-request-handler.ts]: Issue is fixed in latest updates.

**Comment by yongkanm** [comment]
> ### Table of String Formats for Addition
> Here is a table of how the addition description should look.
> 
> #### Adding without optional total points
> Adding `10%`|`10` to assignment with `0.02%`|`1` of `50` possible points.
> 
> % assignment	points assignment
> add %	`10% to 0.02% of 50 points`	`10% of 50 points to 1 out of 50 points`
> add points	`10 to 0.02% of 50 points`	`10 to 1 out of 50 points`
> #### Adding with optional total points
> Adding `10%`|`10` of `200` total assignment group points to assignment with `0.02%`|`1` of `50` possible points.
> 
> % assignment	points assignment
> add %	`10% of 200 total points to 0.02% of 50 points`	`10% of 200 total points to 1 out of 50 points`
> add points	`10 to 0.02% of 50 points`	`10 to 1 out of 50 points`
> Let me know if you think these should be changed.

I think this format is good enough. We might need to change it later upon user request, but currently I cannot think of anything better than this.

**Review by yongkanm — APPROVED** [review]
The following fixes have been tested & work as expected:

1. Add `await` to the posting grade API call in request handler
2. Add negative number support
3. Move the check for grading type after other local guards & add another check for publish state
4. Fix misleading descriptions in the comment

This pull request is ready for merge now.

File: src/app/services/canvas.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/services/canvas.service.ts:1592]: Agreed. Maybe something similar to the mixin pattern would help, but it would be a pain to make the type system work as expected when the combination of multiple extensions are possible.

[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/services/canvas.service.ts:1592]: TypeScript provides a utility type `Partial<T>`, which returns a type by making all properties of `T` optional. The opposite is `Required<T>`. Combining these two utility types with type intersection `&` could achieve what you want. The hard part is that if we have multiple flags (e.g., `hasGradingType` and `hasPublishState`), changing the function return value in correspondance of these flags could be difficult (but I think it's still feasible with TypeScript).
